### PR TITLE
Implemented validation & authz for createUser

### PR DESCRIPTION
**Pull request by simong**

### CODE REVIEW

File: node_modules/gh-core/lib/db.js [hunk 1]
```diff
@@ -215,7 +215,7 @@ var _setUpModel = function(sequelize) {
         'email': {
             'type': Sequelize.STRING,
             'allowNull': true,
-            'unique': true
+            'unique': 'emailUnique'
```
Comment (nicolaasmatthijs) [inline node_modules/gh-core/lib/db.js]: Can you add a line of docs about `emailUnique`

File: node_modules/gh-users/lib/api.js [hunk 2]
```diff
@@ -182,16 +165,32 @@ var getUserByCredentials = module.exports.getUserByCredentials = function(appId,
  */
 var createUser = module.exports.createUser = function(ctx, appId, userProfile, credentials, callback) {
     userProfile = userProfile || {};
+    credentials = credentials || {};
     userProfile.emailPreference = userProfile.emailPreference || 'immediate';
     userProfile.isAdmin = GrasshopperUtil.getBooleanParam(userProfile.isAdmin, false);
```
Comment (nicolaasmatthijs) [inline node_modules/gh-users/lib/api.js]: Should convert app id to integer

File: node_modules/gh-users/lib/api.js [hunk 3]
```diff
@@ -182,16 +165,32 @@ var getUserByCredentials = module.exports.getUserByCredentials = function(appId,
...
-    // TODO: displayName validation
-    // TODO: email validation
-    // TODO: email preference validation
-    // TODO: isAdmin validation
-    // TODO: strategy validation
-    // TODO: validate password when using local
-    // TODO: validate shibbolethid when using shib
+    var validator = new Validator();
+    validator.check(appId, {'code': 400, 'msg': 'A valid app id must be provided'}).isInt();
+    validator.check(userProfile.displayName, {'code': 400, 'msg': 'A valid displayName must be provided'}).notEmpty();
```
Comment (nicolaasmatthijs) [inline node_modules/gh-users/lib/api.js]: Display name has max length as well

File: node_modules/gh-users/lib/api.js [hunk 4]
```diff
@@ -182,16 +165,32 @@ var getUserByCredentials = module.exports.getUserByCredentials = function(appId,
...
+        validator.check(credentials.password, {'code': 400, 'msg': 'A valid password must be provided'}).notEmpty();
+        validator.check(credentials.password, {'code': 400, 'msg': 'A valid password must be provided'}).len(6);
+    } else if (credentials.strategy === 'shibboleth') {
+        validator.check(credentials.shibbolethId, {'code': 400, 'msg': 'A valid shibboleth identifier must be provided'}).notEmpty();
+    }
+    if (validator.hasErrors()) {
+        return callback(validator.getFirstError());
+    }
+
+    var canNotAdminApp = (!ctx.user || !ctx.user.canAdmin(appId));
```
Comment (nicolaasmatthijs) [inline node_modules/gh-users/lib/api.js]: I would find something like `isAppAdmin` a bit easier to read

File: node_modules/gh-users/tests/test-users.js [hunk 5]
```diff
@@ -39,6 +41,173 @@ describe('Users', function() {
                 });
             });
         });
+
+        /**
+         * Verify validation when creating a user
+         */
+        it('verify validation when creating a user', function(callback) {
+            var email = TestsUtil.generateTestEmailAddress();
+            // Missing displayName
```
Comment (nicolaasmatthijs) [inline node_modules/gh-users/tests/test-users.js]: Should do displayName that is too long

File: node_modules/gh-users/tests/test-users.js [hunk 6]
```diff
@@ -39,6 +41,173 @@ describe('Users', function() {
...
+        });
+
+        /**
+         * Verify authorization when creating a user
+         */
+        it('verify authorization when creating a user', function(callback) {
+            // Generate test tenant
+            TestsUtil.generateTestTenant(1, function(tenant, app) {
+
+                // Enable local account creation
```
Comment (nicolaasmatthijs) [inline node_modules/gh-users/tests/test-users.js]: Should we also validate that an admin from tenant A cannot create a user on tenant B ?

File: node_modules/gh-users/tests/util.js [hunk 7]
```diff
@@ -68,6 +68,56 @@ var assertGetMe = module.exports.assertGetMe = function(client, callback) {
...
+ * @param  {String}             [opts.recaptchaChallenge]       The identifier for the recaptcha challenge. Only required when the current user is not an app administrator
+ * @param  {String}             [opts.recaptchaResponse]        The recaptcha response entered for the presented challenge. Only required when the current user is not an app administrator
+ * @param  {Function}           callback                        Standard callback function
+ * @param  {Object}             callback.user                   The basic user profile of the created test user
+ */
+var assertCreateUser = module.exports.assertCreateUser = function(client, displayName, email, password, options, callback) {
+    client.user.createUser(displayName, email, password, options, function(err, user) {
+        assert.ok(!err);
+        assert.ok(user);
+        return callback(user);
```
Comment (nicolaasmatthijs) [inline node_modules/gh-users/tests/util.js]: Should we validate the properties on the created user as well?

**Comment by nicolaasmatthijs** [comment]
Finished review. Back to @simong for follow-up

File: node_modules/gh-users/tests/test-users.js (hunk 6 above)
Comment (simong) [inline node_modules/gh-users/tests/test-users.js]: App admins can't specify an app id, users just get created on their own application

**Comment by simong** [comment]
I totally forgot to push my follow-up. Oops!

**Comment by nicolaasmatthijs** [comment]
Left some final review. Back to @simong for follow-up

**Comment by simong** [comment]
Done. I rebased this branch against master to bring in some small fixes for sequelize though. Back to @nicolaasmatthijs

**Comment by nicolaasmatthijs** [comment]
Looks good. Merged.

**Pull request merged**